[JIT] simplify logic for merging types during profiling

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69096

Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

Differential Revision: [D32772626](https://our.internmc.facebook.com/intern/diff/D32772626)

davidberard98 committed:
[WIP][JIT] simplify logic for merging types during profiling

Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

swolchok commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
-      if (t.defined()) {
-        auto pttp = tensorTypeInCurrentExecutionContext(t);
+      auto new_tensor_type =
+          (t.defined() ? tensorTypeInCurrentExecutionContext(t)
+                       : TensorType::get()->withUndefined());
```
doesn't tensorTypeInCurrentExecutionContext already do this?

swolchok commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
+          *new_tensor_type);
+
+      if (pn->hasRun()) {
+        auto existing_tensor_type =
+            pn->ty(attr::profiled_type)->expect<TensorType>();
```
You don't need the shared_ptr, so use expectRef and save a reference count bump
`const auto& existing_tensor_type = pn->ty(attr::profiled_type)->expectRef<TensorType>();`

swolchok commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
+            *existing_tensor_type);
+        auto merged_type = new_tensor_type->merge(*existing_tensor_type);
+        GRAPH_DEBUG("Merged type for %", pno->debugName(), ": ", *merged_type);
+        pn->ty_(attr::profiled_type, merged_type);
```
save a refcount bump here. would be slightly faster still to inline the call when the GRAPH_DEBUG won't fire, but we probably don't need to go there until profiling data says we do.
`         pn->ty_(attr::profiled_type, std::move(merged_type));`

swolchok commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
       } else {
-        profiled_types[pno] = TensorType::get()->withUndefined();
+        pn->setHasRun(true);
+        pn->ty_(attr::profiled_type, new_tensor_type);
```
same, missing move

`pn->ty(attr::profiled_type, std::move(new_tensor_type));`

swolchok commented on torch/csrc/jit/runtime/profiling_record.h:
```diff
@@ -184,15 +184,11 @@ struct ProfilingRecord {
-  // the key is a frame id
-  // the value is a mapping from a Value in a graph
-  // to a profiled TensorType
-  std::map<int64_t, std::map<Value*, TensorTypePtr>> profiled_types_per_frame_;
```
cool, no need to use std::unordered_map or F14 if the map is just gone!

swolchok commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
-            " with ",
-            *pttp);
-        if (profiled_types.count(pno) == 0) {
-          profiled_types.insert({pno, pttp});
```
this had a missing move, so if we ever need it back we should put the move in

davidberard98 committed:
Update on "[WIP][JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

Krovatkin commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -144,35 +144,42 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
-          profiled_types.insert({pno, pttp});
+            *new_tensor_type);
+
+        if (pn->hasRun()) {
```
could we use `hasAttr(attr::profiled_type)` on `pn`?

davidberard98 replied:
@Krovatkin there's a few places that expect `attr::profiled_type` to be set; I tried removing the initialization (i.e. `pn->ty_(attr::profiled_type, TensorType::get());`) but a bunch of tests fail, e.g.

```
ERROR: test_bias_as_arg (jit.test_autodiff_subgraph_slicing.TestAutodiffSubgraphSlicing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/dberard/pytorch/test/jit/test_autodiff_subgraph_slicing.py", line 135, in test_bias_as_arg
    check_against_reference(self, scripted, method1, lambda x: x, (x, weight, bias), check_types=False)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 128, in check_against_reference
    grads_test = torch.autograd.grad(
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/autograd/__init__.py", line 275, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: required keyword attribute 'profiled_type' is undefined

----------------------------------------------------------------------
Ran 26 tests in 2.127s

FAILED (errors=1)
```

Krovatkin replied:
@davidberard98 could you please elaborate why switching 
```
if (pn->hasRun() {
// ... merge types
} else {
 pn->ty_(attr::profiled_type, new_incoming_type);
}
```

to 

```
if (pn->hasAttribute(attr::profiled_type)) {
// ... merge types
} else {
 pn->ty_(attr::profiled_type, new_incoming_type);
 // next time we call `hasAttribute` we merge types
} 
```

davidberard98 replied:
@Krovatkin 
Currently, we initialize `pn->ty_(attr::profiled_type, TensorType::get())` before the lambda gets called (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L145). So if we simply replaced `hasRun()` with `hasAttribute()`, then `pn->hasAttribute(attr::profiled_type)` would always evaluate to true.

If we remove the initialization, then we get the `required keyword attribute 'profiled_type' is undefined` error that I mentioned.

Krovatkin commented on torch/csrc/jit/runtime/profiling_record.h:
```diff
@@ -184,15 +184,11 @@ struct ProfilingRecord {
   // `new_sizes` and `sym_shapes` are of the same rank
 
-  c10::SymbolicShape mergeSymbolicShapes(
+  static c10::SymbolicShape mergeSymbolicShapes(
```
I was hoping we can nuke this helper? Are there any uses?

davidberard98 replied:
whoops, missed that - I think you're right

Krovatkin reviewed (commented):
This looks amazing! A few quick questions!

davidberard98 committed:
Update on "[JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

Differential Revision: [D32772626](https://our.internmc.facebook.com/intern/diff/D32772626)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

Krovatkin reviewed (approved):
:shipit:

davidberard98 committed:
Update on "[JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

Differential Revision: [D32772626](https://our.internmc.facebook.com/intern/diff/D32772626)

[ghstack-poisoned]

Krovatkin commented:
@swolchok do you want to do another pass or can we merge this?

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

davidberard98 committed:
Update on "[JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

Differential Revision: [D32772626](https://our.internmc.facebook.com/intern/diff/D32772626)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

swolchok commented:
> @swolchok do you want to do another pass or can we merge this?

no complaints from me!

davidberard98 committed:
Update on "[JIT] simplify logic for merging types during profiling"


Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

Differential Revision: [D32772626](https://our.internmc.facebook.com/intern/diff/D32772626)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).
